exchanges/subscription/subscription.go

```diff
@@ -1,92 +1,144 @@
-	Authenticated bool                   `json:"authenticated,omitempty"`
+	Enabled       bool           `json:"enabled"`
+	Key           any            `json:"-"`
+	Channel       string         `json:"channel,omitempty"`
+	Pairs         currency.Pairs `json:"pairs,omitempty"`
+	Asset         asset.Item     `json:"asset,omitempty"`
+	Params        map[string]any `json:"params,omitempty"`
+	Interval      kline.Interval `json:"interval,omitempty"`
+	Levels        int            `json:"levels,omitempty"`
+	Authenticated bool           `json:"authenticated,omitempty"`
+	state         State
+	m             sync.RWMutex
 }
 
-// MarshalJSON generates a JSON representation of a Subscription, specifically for config writing
-// The only reason it exists is to avoid having to make Pair a pointer, since that would be generally painful
-// If Pair becomes a pointer, this method is redundant and should be removed
-func (s *Subscription) MarshalJSON() ([]byte, error) {
-	// None of the usual type embedding tricks seem to work for not emitting an nil Pair
-	// The embedded type's Pair always fills the empty value
-	type MaybePair struct {
-		Enabled       bool                   `json:"enabled"`
-		Channel       string                 `json:"channel,omitempty"`
-		Asset         asset.Item             `json:"asset,omitempty"`
-		Params        map[string]interface{} `json:"params,omitempty"`
-		Interval      kline.Interval         `json:"interval,omitempty"`
-		Levels        int                    `json:"levels,omitempty"`
-		Authenticated bool                   `json:"authenticated,omitempty"`
-		Pair          *currency.Pair         `json:"pair,omitempty"`
-	}
+// String implements the Stringer interface for Subscription, giving a human representation of the subscription
+func (s *Subscription) String() string {
+	p := s.Pairs.Format(currency.PairFormat{Uppercase: true, Delimiter: "/"})
+	return fmt.Sprintf("%s %s %s", s.Channel, s.Asset, p.Join())
+}
 
-	k := MaybePair{s.Enabled, s.Channel, s.Asset, s.Params, s.Interval, s.Levels, s.Authenticated, nil}
-	if s.Pair != currency.EMPTYPAIR {
-		k.Pair = &s.Pair
-	}
+// State returns the subscription state
+func (s *Subscription) State() State {
+	s.m.RLock()
+	defer s.m.RUnlock()
+	return s.state
+}
 
-	return json.Marshal(k)
+// SetState sets the subscription state
+// Errors if already in that state or the new state is not valid
+func (s *Subscription) SetState(state State) error {
+	s.m.Lock()
+	defer s.m.Unlock()
+	if state == s.state {
+		return ErrInStateAlready
+	}
+	if state > UnsubscribedState {
+		return ErrInvalidState
+	}
+	s.state = state
+	return nil
 }
 
-// String implements the Stringer interface for Subscription, giving a human representation of the subscription
-func (s *Subscription) String() string {
-	return fmt.Sprintf("%s %s %s", s.Channel, s.Asset, s.Pair)
+// SetKey does what it says on the tin safely for concurrency
+func (s *Subscription) SetKey(key any) {
```

Comment:
You mean nil check on the Key, correct?
I think it's valid to set the key to nil, and I've done it on Clone() subs.
Since the key is any, I think it's fine.